File: packages/ios/Sources/RichTextiOS/AST.swift
import Foundation

public enum NodeType: String, Equatable, Sendable, CaseIterable {
    case document
    case paragraph
    case heading
    case codeBlock
    case list
    case listItem
    case text
    case bold
    case italic
    case inlineCode
    case link
    case mention
    case hardBreak
}

public struct MentionPayload: Equatable, Sendable, Hashable {
    public var userId: String
    public var userName: String

    public init(userId: String, userName: String) {
        self.userId = userId
        self.userName = userName
    }
}

public enum InlineNode: Equatable, Sendable {
    case text(String)
    case mention(MentionPayload)
    case hardBreak
    case bold([InlineNode])
    case italic([InlineNode])
    case inlineCode(String)
    case link(href: String, children: [InlineNode])
}

public enum BlockNode: Equatable, Sendable {
    case paragraph([InlineNode])
    case heading(level: Int, children: [InlineNode])
    case codeBlock(language: String?, value: String)
    case list(ordered: Bool, start: Int?, items: [ListItemNode])
}

public struct ListItemNode: Equatable, Sendable {
    public var children: [ListItemChild]

    public init(children: [ListItemChild]) {
        self.children = children
    }
}

public enum ListItemChild: Equatable, Sendable {
    case paragraph([InlineNode])
    case list(ordered: Bool, start: Int?, items: [ListItemNode])
}

public struct DocumentNode: Equatable, Sendable {
    public var children: [BlockNode]

    public init(children: [BlockNode] = []) {
        self.children = children
    }
}

File: packages/android/richtext-android/src/main/java/com/company/richtext/ast/Ast.kt
package com.company.richtext.ast

enum class NodeType {
    DOCUMENT,
    PARAGRAPH,
    HEADING,
    CODE_BLOCK,
    LIST,
    LIST_ITEM,
    TEXT,
    BOLD,
    ITALIC,
    INLINE_CODE,
    LINK,
    MENTION,
    HARD_BREAK,
}

data class MentionPayload(
    val userId: String,
    val userName: String,
)

sealed interface InlineNode {
    data class Text(val value: String) : InlineNode
    data class Mention(val payload: MentionPayload) : InlineNode
    data object HardBreak : InlineNode
    data class Bold(val children: List<InlineNode>) : InlineNode
    data class Italic(val children: List<InlineNode>) : InlineNode
    data class InlineCode(val value: String) : InlineNode
    data class Link(val href: String, val children: List<InlineNode>) : InlineNode
}

sealed interface BlockNode {
    data class Paragraph(val children: List<InlineNode>) : BlockNode
    data class Heading(val level: Int, val children: List<InlineNode>) : BlockNode
    data class CodeBlock(val language: String?, val value: String) : BlockNode
    data class ListBlock(
        val ordered: Boolean,
        val start: Int?,
        val items: List<ListItemNode>,
    ) : BlockNode
}

data class ListItemNode(
    val children: List<ListItemChild>,
)

sealed interface ListItemChild {
    data class Paragraph(val children: List<InlineNode>) : ListItemChild
    data class NestedList(
        val ordered: Boolean,
        val start: Int?,
        val items: List<ListItemNode>,
    ) : ListItemChild
}

data class DocumentNode(
    val children: List<BlockNode> = emptyList(),
)
